### PR TITLE
Add option to trigger outgoing webhook if first word starts with trigger word

### PR DESCRIPTION
**Pull request by Rudloff**
#### Summary
This patch adds a new option when adding an outgoing webhook:
mattermost webhook
The first option is the current behavior.
The second one triggers the hook if the first word of a post starts with the trigger word (e.g if your trigger word is _foo_, _foobar_ will trigger it).

#### Ticket Link
https://github.com/mattermost/platform/issues/3574

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] Has driver changes that have been merged and package.json updated
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [ ] Includes text changes and localization files updated
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by jwilander** [comment]
@jasonblais can you take a look?

**Comment by jasonblais** [comment]
Thanks @Rudloff for your submission! Love your initiative here

Sorry for the delay in the review process, currently testing different cases and will get back to you on this tomorrow.

Thanks again!!

**Comment by jasonblais** [comment]
Thanks @Rudloff once more -- I was having technical difficulties getting one of our webhooks set up, but it's hooked now and I was able to test it successfully. Thank you for your patience -- I'll be able to test future updates on this much sooner.

I tested your pull request on Windows 10, Chrome, and set up outgoing webhooks using a [Giphy Integration Service for Mattermost](https://github.com/numberly/mattermost-integration-giphy). 

1) If I select `Trigger when: first word matches trigger word exactly`, everything works as expected

2) If I select `Trigger when: first word starts with trigger word`, with trigger word `gif`
 - sending a message `gif hello` sends a response and I receive a giphy as expected
 - sending a message `gifhello hello`, `giff hello` or other combinations provides no results

Hence, it appears I'm unable to get this to work as expected? I've attached a screenshot of two webhooks I've tried. You may also use the test server above, if you'd like to try it out.

image

3) Will this change apply if I have specified multiple trigger words? For instance, if I had `gif` and `giphy`, with the option `Trigger when first word starts with trigger word`, will posts starting with EITHER `gif` OR `giphy` trigger the webhook?

4) For the UI text, propose making slight tweaks for it:
 - `Trigger when` --> `Trigger When`
 - `First word matches trigger word exactly` --> `First word matches a trigger word exactly`
 - `First word starts with trigger word` --> `First word starts with a trigger word`

image

**Comment by Rudloff** [comment]
Sorry, it looks like I made a mistake with my last push and some of the code was missing :/
I updated my PR and it should work correctly now.
(I also updated the strings as you suggested.)

To answer 3: yes it should work with either word if you add multiple trigger words.

**Comment by jasonblais** [comment]
+1 for PM approval

Tested the following cases successfully on Windows 10, Chrome (trigger words contain gif)

 - messages triggering webhook (each one followed with the word 'hey'): gif, gif#, gif-hey, gifhey, gif*, gif:, gif//
 - messages not triggering the webhook: #gif, agif, .gif, # gif, hey gif

Tested above cases in the following situations:
 - when only one trigger word specified, as well as when multiple trigger words are specified.
 - when `First word matches a trigger word exactly` and when `First word starts with a trigger word` options chosen

No issues found, big thank you @Rudloff! This is a nice improvement for outgoing webhooks :+1:

**Comment by enahum** [comment]
Hi @Rudloff thanks for the PR, can you please add the relevant text to the correspondent `en.json` files?

**Comment by Rudloff** [comment]
@enahum I updated my PR. Is it OK like this?

File: webapp/components/integrations/components/add_outgoing_webhook.jsx [hunk 1]
```diff
@@ -300,6 +309,41 @@ export default class AddOutgoingWebhook extends React.Component {
                         <div className='form-group'>
                             <label
                                 className='control-label col-sm-4'
+                                htmlFor='triggerWords'
+                            >
+                                <FormattedMessage
+                                    id='add_outgoing_webhook.triggerWordsTriggerWhen'
```
Comment (enahum) [inline webapp/components/integrations/components/add_outgoing_webhook.jsx]: Thanks @Rudloff for your quick response.

This one is still missing in `en.json` my friend.

Comment (Rudloff) [inline webapp/components/integrations/components/add_outgoing_webhook.jsx]: Oh sorry, I forgot it. I updated the PR.

**Comment by enahum** [comment]
Thanks +1

File: api/post.go [hunk 2]
```diff
@@ -384,11 +384,19 @@ func handleWebhookEvents(c *Context, post *model.Post, team *model.Team, channel
 	relevantHooks := []*model.OutgoingWebhook{}
 	for _, hook := range hooks {
 		if hook.ChannelId == post.ChannelId {
-			if len(hook.TriggerWords) == 0 || hook.HasTriggerWord(firstWord) {
+			if len(hook.TriggerWords) == 0 {
+				relevantHooks = append(relevantHooks, hook)
+			} else if hook.TriggerWhen == 0 && hook.HasTriggerWord(firstWord) {
```
Comment (crspeller) [inline api/post.go]: Can we get some consents for these values? Maybe have a function that encapsulates this logic? This whole section is really hard to understand.

File: model/outgoing_webhook.go [hunk 3]
```diff
@@ -204,3 +210,17 @@ func (o *OutgoingWebhook) HasTriggerWord(word string) bool {
 
 	return false
 }
+
+func (o *OutgoingWebhook) TriggerWordStartsWith(word string) bool {
```
Comment (crspeller) [inline model/outgoing_webhook.go]: Unit test?

Comment (Rudloff) [inline model/outgoing_webhook.go]: I added a basic unit test. Is that enough?

File: api/post.go (hunk 2 above)
Comment (Rudloff) [inline api/post.go]: I think we can refactor like that:
```go
for _, hook := range hooks {
	if hook.ChannelId == post.ChannelId || len(hook.ChannelId) == 0 {
		if hook.ChannelId == post.ChannelId && len(hook.TriggerWords) == 0 {
			relevantHooks = append(relevantHooks, hook)
		} else if hook.TriggerWhen == 0 && hook.HasTriggerWord(firstWord) {
			relevantHooks = append(relevantHooks, hook)
		} else if hook.TriggerWhen == 1 && hook.TriggerWordStartsWith(firstWord) {
			relevantHooks = append(relevantHooks, hook)
		}
	}
}
```
Would that be OK for you?

Comment (crspeller) [inline api/post.go]: Better, but let's get some constants for the TriggerWhen `0` and `1` values.

File: api/post.go [hunk 4]
```diff
@@ -383,12 +388,14 @@ func handleWebhookEvents(c *Context, post *model.Post, team *model.Team, channel
 
 	relevantHooks := []*model.OutgoingWebhook{}
 	for _, hook := range hooks {
```
Comment (Rudloff) [inline api/post.go]: @crspeller Is it OK like this?

Comment (crspeller) [inline api/post.go]: @Rudloff Yes, after you fix the gofmt error.

Comment (Rudloff) [inline api/post.go]: Whoops, sorry I forgot to run `make check-server-style`.
It should be good now.

**Comment by crspeller** [comment]
+1

**Pull request merged**